doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,465 @@
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics, and allow components
+  of a system to each work with a different schema version and merge
+  their results.
+- Hosts and managers would be capable of working with mismatched
+  schema versions, so long as the versions of the individual required
+  traits are compatible.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+- Managers not required to consider storage/communication of additional
+  version data to enable interpretation by future/downstream systems.
+- Stateless managers remain possible
+- Allows per-trait versioning, which avoids the need or "insider
+  knowledge" as to which traits actually change in which package
+  versions.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would be become a tedious dance.
+- Introspection methods become more complex. (See [What's
+  versioned](#whats-versioned)).
+- Unclear how specification versioning is to work without an out-of-data
+  version.
+- The most obvious implementation, (using `traitID` to store the
+  version), would effectively make any change to a trait version a
+  breaking change. Allowing non breaking changes (such as adding a new
+  property) would require significant rework to the `traitgen` output.
+
+### In data - Extended
```

Comment:
[consider] I wonder if we really need this section, since its essentially saying that introspect/negotiation can be layered on, which is a Pro, but the same applies to the out-of-data approach (i.e. in-data could be layered on top of negotiation at a later date), so doesn't really differentiate the options in any way.

I think it's worth adding a little note in the "Rationale" section that choosing this option doesn't preclude layering on introspection or negotiation at a later date, i.e. the approach is extensible.

Reply:
I know it's a _tad_ awkward, but I think it's a neccesary callout as a large part of the rationale is due to how in-data is more naturally extensible and fits into the openassetio system better.

That is already called out in the rationale. 

>  a large reason for the adoption of an in-data solution is
how little of the API it needs to disrupt in order to become useful, and
yet how open it is to extension.

and 

> ...allows a layered
iterative approach to reduce this complexity, rather than having to do a
big bang drop full of speculation all at once. We look forward with
excited expectation as to how `openassetio-traitgen`, with its generated
views, can be leveraged to make this experience simpler and easier.

Reply:
The rationale section doesn't explicitly mention layering on introspection/negotiation as a possibility. But if we're keeping this "Extended" section, then fine.